packages/test/test-end-to-end-tests/src/mocking.ts

```diff
@@ -0,0 +1,63 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { isPromiseLike } from "@fluidframework/core-utils";
+
+export type UnPromise<T> = T extends Promise<infer U> ? U : T;
+
+export type OverrideFactory<T, P extends keyof T> = (T: T) => T[P];
+
+type NestedOverrides<T> = {
+	[P in keyof T]?: T[P] extends (...args: any) => any
+		? NestedOverrides<UnPromise<ReturnType<T[P]>>> | OverrideFactory<T, P>
+		: OverrideFactory<T, P>;
+};
+
+export function wrapObjectAndOverride<T extends Record<string, any>>(
```

Comment:
Did you look into Sinon?  https://sinonjs.org/releases/v17/  It's very possible it offers similar functionality out of the box.

Reply:
I'll admit I have also found it difficult to understand/use at times, this API looks potentially more approachable.  Consider using Sinon to implement it though.

Reply:
briefly, and didn't see anything that looked similar. we already had this is our code base, just reusing it.

Reply:
i didn't add this in this PR, just copy pasted from packages/test/test-end-to-end-tests/src/test/batchBreakRegression.spec.ts